### PR TITLE
User active

### PR DESCRIPTION
**Pull request by krulis-martin**

### CODE REVIEW

File: app/commands/RemoveInactiveUsers.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    $this->anonymizationHelper = $anonymizationHelper;
+  }
+
+  protected function configure() {
+    $this->setName('users:remove-inactive')->setDescription('Remove users who has not been active for some time.');
+    $this->addOption('silent', null, InputOption::VALUE_NONE, 'Just do the job (no interactions, no reports)');
+  }
+
+  protected function execute(InputInterface $input, OutputInterface $output) {
+    if (!$this->inactivityThreshold) return;
```
Comment (Neloop) [inline app/commands/RemoveInactiveUsers.php]: PSR-2! :-)

File: app/commands/RemoveInactiveUsers.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      if (!$helper->ask($input, $output, $question)) {
+        $output->writeln("Aborted.");
+        return;
+      }
+    }
+
+    // perform deletion
+    foreach ($usersToDelete as $user) {
+      $this->anonymizationHelper->prepareUserForSoftDelete($user);
+      $this->users->remove($user);
```
Comment (Neloop) [inline app/commands/RemoveInactiveUsers.php]: Just curious, is there some method for entity deletion which handles array of entities?

File: app/config/config.neon [hunk 3]
```diff
@@ -148,6 +148,13 @@ parameters:
       solutionPointsUpdatedPrefix: "Assignment Points Updated - "
       shadowPointsUpdatedPrefix: "Shadow Assignment Points Awarded - "
 
+  generalStatsNotifications:
+    period: "1 week"
+    emails:
+      from: %emails.from%
+      to: %failures.emails.to%
```
Comment (Neloop) [inline app/config/config.neon]: This is rather unexpected dependency on other specific configuration parameter, I would suggest to establish new general parameter with name for example `emails.adminEmail` or similar and use this one in `failures` and in here.

File: app/config/config.neon [hunk 4]
```diff
@@ -191,6 +198,11 @@ parameters:
     anonymizedName: "@@anonymized"  # String used as replacement for anonymized name a user
     deletedEmailSuffix: "@@deleted.recodex"  # Suffix string appended to an email address of a user, when account is deleted
 
+  removeInactiveUsers:
+    # How long the user has to be inactive to warant the removal (null = never remove students, 1 month is minimum).
+    # Please note that the length of the auth. token expiration should be considered (readonly tokens may expire after 1 year).
```
Comment (Neloop) [inline app/config/config.neon]: 👍

File: app/helpers/Emails/Notifications/SubmissionEmailsSender.php [hunk 5]
```diff
@@ -37,7 +37,7 @@ class SubmissionEmailsSender {
   public function __construct(EmailHelper $emailHelper, array $params) {
     $this->emailHelper = $emailHelper;
     $this->sender = Arrays::get($params, ["emails", "from"], "noreply@recodex.mff.cuni.cz");
-    $this->submissionEvaluatedPrefix = Arrays::get($params, ["emails", "submissionEvaluatedPrefix"], "Submission Evaluated - ");
+    $this->submissionEvaluatedPrefix = Arrays::get($params, ["emails", "subject"], "Submission Evaluated - ");
```
Comment (Neloop) [inline app/helpers/Emails/Notifications/SubmissionEmailsSender.php]: This feels like unintended change and victim of find-all replace-all and before that copy&paste :-)

File: app/helpers/GeneralStatsHelper.php [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+ */
+class GeneralStatsHelper {
+  use Nette\SmartObject;
+
+  /** @var string */
+  private $inactivityThreshold;
+
+  /**
+   * @var Users
+   * @inject
```
Comment (Neloop) [inline app/helpers/GeneralStatsHelper.php]: Inject? Really?

File: app/helpers/GeneralStatsHelper.php [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+   */
+  public $users;
```
Comment (Neloop) [inline app/helpers/GeneralStatsHelper.php]: Public? Really?

File: app/model/repository/BaseRepository.php [hunk 8]
```diff
@@ -80,6 +101,10 @@ public function matching(Criteria $params) {
     return $this->repository->matching($params);
   }
 
+  public function createQueryBuilder(string $alias, string $indexBy = null)
```
Comment (Neloop) [inline app/model/repository/BaseRepository.php]: Is visibility `public` really on point here?

File: app/model/repository/ReferenceSolutionSubmissions.php [hunk 9]
```diff
@@ -20,4 +14,20 @@ public function __construct(EntityManager $em) {
...
   }
 
+  /**
+   * Find all submissions created in given time interval.
+   * @param DateTime|null $since Only submissions created after this date are returned.
+   * @param DateTime|null $until Only submissions created before this date are returned.
+   */
+  public function findByCreatedAt(?DateTime $since, ?DateTime $until)
+  {
+    $qb = $this->createQueryBuilder('s'); // takes care of softdelete cases
```
Comment (Neloop) [inline app/model/repository/ReferenceSolutionSubmissions.php]: You wrote similar function three times in this PR, maybe probably some helper function on `BaseRepository` can be introduced?

File: app/commands/RemoveInactiveUsers.php (hunk 1 above)
Comment (krulis-martin) [reply app/commands/RemoveInactiveUsers.php]: Yeah, we do not break lines at the beginning of function and class declarations (which is one of the best things in PSR, but you choose to ignore it), but let's badger poor ol' me for not spreading a simple guard over 3 lines.

File: app/commands/RemoveInactiveUsers.php (hunk 2 above)
Comment (krulis-martin) [reply app/commands/RemoveInactiveUsers.php]: I do not think so.

File: app/helpers/GeneralStatsHelper.php (hunk 6 above)
Comment (krulis-martin) [reply app/helpers/GeneralStatsHelper.php]: Just copypasting your code :)

Comment (Neloop) [reply app/helpers/GeneralStatsHelper.php]: Where from? In case this was in some helper, I shall be called idiot then...

Comment (krulis-martin) [reply app/helpers/GeneralStatsHelper.php]: OK, maybe its not your code. I shall try to find and fix it at source as well.

**Pull request merged**